### PR TITLE
fix for LOG-2159 Cluster Logging Pods in CrashLoopBackOff

### PR DESCRIPTION
**Pull request by pmoogi-redhat**
### Description
patch for fixing fluentd issue on the use of yajl json parser. 

/cc @jcantrill 

### Links
- JIRA: https://issues.redhat.com/browse/LOG-2159

### CODE REVIEW

File: fluentd-fix-LOG-2159.patch [hunk 1]
```diff
@@ -0,0 +1,22 @@
+diff --git a/fluentd/vendored_gem_src/fluentd/lib/fluent/plugin/parser_json.rb b/fluentd/vendored_gem_src/fluentd/lib/fluent/plugin/parser_json.rb
+index 7be12efa..ca8e67fb 100644
+--- a/fluentd/vendored_gem_src/fluentd/lib/fluent/plugin/parser_json.rb
++++ b/fluentd/vendored_gem_src/fluentd/lib/fluent/plugin/parser_json.rb
+@@ -54,12 +54,15 @@ module Fluent
+           Oj.default_options = Fluent::DEFAULT_OJ_OPTIONS
+           [Oj.method(:load), Oj::ParseError]
+         when :json then [JSON.method(:load), JSON::ParserError]
+-        when :yajl then [Yajl.method(:load), Yajl::ParseError]
++        #where is no oj parser, json parser falls back to yajl parser and it is throwing errors on certain platforms like Z-series. As a fix disabling it completely
```
Comment (jcantrill) [inline fluentd-fix-LOG-2159.patch:10]: I'm not certain this is the right approach.  Why are these parser's not available on Z?  Do C platforms have standard C libraries that are not available to us?  How does the runtime change that is different from build time? The native bits of each gem are configured here:

* https://github.com/ohler55/oj/blob/develop/ext/oj/extconf.rb
* https://github.com/brianmario/yajl-ruby/blob/master/ext/yajl/extconf.rb

We build these images to run on a RHEL base image.  My understanding is, from container perspective, all its dependencies should be and run on RHEL.  Why does the underlying platform being Z matter?  Additionally, why does the default JSON parser even change?

@syedriko  help me understand

Comment (syedriko) [reply fluentd-fix-LOG-2159.patch:10]: +1 This entire story makes no sense to me. We install the same gems and build their native parts across all supported platforms.

Comment (pmoogi-redhat) [reply fluentd-fix-LOG-2159.patch:10]: I didn't have z-cluster to debug it further @jcantrill and @syedriko. will go deeper to answer your question. one thing for sure we didn't have this issue raised on any of the AWS cluster.

Comment (pmoogi-redhat) [reply fluentd-fix-LOG-2159.patch:10]: @jcantrill and @syedriko worked with QE folks today. Below is confirmed and checked with their help.

$ oc rsh fluentd-qzzsd
sh-4.4# ls -ltR | grep -i yajl
sh-4.4# cd /usr/local/share/gems/gems/
sh-4.4# ls | grep -iE 'yajl|oj'
oj-3.11.3
yajl-ruby-1.4.1

So oj and yajl parser libs installed part of gem as expected

$ oc debug node/ikanse-11-gwgkm-worker-c-78n5t.c.openshift-qe.internal
sh-4.4# chroot /host
sh-4.4# cat /var/lib/fluentd/pos/journal_pos.json
{"journal":s=d728b81815b0404aa886b55bb4c9fb6a;i=3307;b=56fee32a91964facb84bdc776cc11821;m=cf5bdb49;t=5d94e41c847ea;x=88ac3e925c5745e5"}

not on Z-cluster as QE mentioned earlier but on Power-cluster if a node crashes for some reason and there is a half-ay written /var/lib/fluentd/pos/journal_pos.json position file as the below:

{"journal":s=d728b81815b0404aa886b55bb4c9fb6a;i=3307;b=56fee32a91964facb84bdc776cc11821;m=cf5bdb49;t=5d94e41c847ea;

then we can reproduce this CrashLoopBackOff problem.

We tried with empty pos file also in that case fluentd is able to handle it.

We saw only a specific node having some issues and got restarted and that led to fluentd CrashLoopBackOff problem.

Now this boils down to how we handle corrupted jason files error handling at fluentd level or oj/yajl parser level.

**Comment by jcantrill** [comment]
@pmoogi-redhat can we close this PR in preference to modifying the fluent run script to remove the position file when corrupted?

**Comment by jcantrill** [comment]
Closing this PR as the issue is closed

**Pull request closed without merging**